### PR TITLE
Dialogue channels internally refresh

### PR DESCRIPTION
**Pull request by carterkozak**
==COMMIT_MSG==
Dialogue channels internally refresh
==COMMIT_MSG==

Quick proof of concept branch, entirely untested

### CODE REVIEW

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 1]
```diff
@@ -495,16 +476,25 @@ private Refreshable<InternalDialogueChannelConfiguration> configurationForServic
...
     }
 
-    @SuppressWarnings("checkstyle:CyclomaticComplexity")
     private ImmutableList<TargetUri> getTargetUris(
             @Safe String serviceNameForLogging,
             Collection<String> uris,
             ProxySelector proxySelector,
             Optional<ImmutableSetMultimap<String, InetAddress>> resolvedHosts) {
+        return getTargetUris(serviceNameForLogging, uris, proxySelector, resolvedHosts, params.taggedMetrics());
+    }
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: The old backcompat version (private) should be removed

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 2]
```diff
@@ -495,16 +476,25 @@ private Refreshable<InternalDialogueChannelConfiguration> configurationForServic
...
+
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    static ImmutableList<TargetUri> getTargetUris(
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: This should be extracted to either DnsSupport or ChannelCache if possible

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 3]
```diff
@@ -543,48 +533,37 @@ private ImmutableList<TargetUri> getTargetUris(
         return ImmutableSet.copyOf(targetUris).asList();
     }
 
-    private static ProxySelector proxySelector(Optional<ProxyConfiguration> proxyConfiguration) {
+    static ProxySelector proxySelector(Optional<ProxyConfiguration> proxyConfiguration) {
         return proxyConfiguration.map(ClientConfigurations::createProxySelector).orElseGet(ProxySelector::getDefault);
     }
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: should be extracted as well

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 4]
```diff
@@ -543,48 +533,37 @@ private ImmutableList<TargetUri> getTargetUris(
...
 
     private Refreshable<InternalDialogueChannel> getInternalDialogueChannel(String serviceName) {
         Preconditions.checkNotNull(serviceName, "serviceName");
         String channelName = ChannelNames.reloading(serviceName, params);
 
-        return configurationForService(serviceName).map(conf -> {
-            Preconditions.checkNotNull(
-                    conf, "Refreshable must not provide a null InternalDialogueChannelConfiguration");
+        return configurationForService(serviceName)
+                .map(InternalDialogueChannelConfiguration::serviceConfiguration)
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: This is a hack, we're doing dns resolution, then throwing away the dns resolution results to get back the original ServiceConfiguration. The DNS piece is entirely unnecessary here!

Comment (carterkozak) [reply dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: resolved

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 5]
```diff
@@ -762,8 +596,5 @@ public String toString() {
     interface InternalDialogueChannelConfiguration {
```
Comment (bjlaub) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: minor nit: do we need this type anymore?

Comment (carterkozak) [reply dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: Good call, might as well remove it!

File: dialogue-clients/src/test/java/com/palantir/dialogue/clients/DialogueClientsDnsIntegrationTest.java [hunk 6]
```diff
@@ -190,7 +193,7 @@ void dns_refresh_updates_in_flight_requests_reloading_channel() throws UnknownHo
         dnsEntries.put(hostOne, InetAddress.getByAddress(hostOne, new byte[] {127, 0, 0, 2}));
 
         TaggedMetricRegistry metrics = new DefaultTaggedMetricRegistry();
-        Counter activeTasks = ClientDnsMetrics.of(metrics).tasks(DnsPollingSpec.RELOADING_FACTORY.kind());
+        Counter activeTasks = ClientDnsMetrics.of(metrics).tasks(ChannelNames.reloading("foo"));
```
Comment (bjlaub) [inline dialogue-clients/src/test/java/com/palantir/dialogue/clients/DialogueClientsDnsIntegrationTest.java:196]: do we want to update these tests to us a higher retry interval + Awaitility instead? Could save that for a separate PR also.

**Review by bjlaub — DISMISSED** [review]

**Pull request merged**